// konva/cmj/shapes/Arc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Arc = void 0;
const Factory_1 = require("../Factory");
const Shape_1 = require("../Shape");
const Global_1 = require("../Global");
const Validators_1 = require("../Validators");
const Global_2 = require("../Global");
class Arc extends Shape_1.Shape {
    _sceneFunc(context) {
        var angle = Global_1.Konva.getAngle(this.angle()), clockwise = this.clockwise();
        context.beginPath();
        context.arc(0, 0, this.outerRadius(), 0, angle, clockwise);
        context.arc(0, 0, this.innerRadius(), angle, 0, !clockwise);
        context.closePath();
        context.fillStrokeShape(this);
    }
    getWidth() {
        return this.outerRadius() * 2;
    }
    getHeight() {
        return this.outerRadius() * 2;
    }
    setWidth(width) {
        this.outerRadius(width / 2);
    }
    setHeight(height) {
        this.outerRadius(height / 2);
    }
    getSelfRect() {
        const radius = this.outerRadius();
        const DEG_TO_RAD = Math.PI / 180;
        const angle = this.angle() * DEG_TO_RAD;
        const inc = 1 * DEG_TO_RAD;
        let start = 0;
        let end = angle + inc;
        if (this.clockwise()) {
            start = end;
            end = 360;
        }
        const xs = [];
        const ys = [];
        for (let i = 0; i < end; i += inc) {
            xs.push(Math.cos(i));
            ys.push(Math.sin(i));
        }
        const minX = Math.round(radius * Math.min(...xs));
        const maxX = Math.round(radius * Math.max(...xs));
        const minY = Math.round(radius * Math.min(...ys));
        const maxY = Math.round(radius * Math.max(...ys));
        return {
            x: minX || 0,
            y: minY || 0,
            width: maxX - minX,
            height: maxY - minY
        };
    }
}
exports.Arc = Arc;
Arc.prototype._centroid = true;
Arc.prototype.className = 'Arc';
Arc.prototype._attrsAffectingSize = ['innerRadius', 'outerRadius'];
(0, Global_2._registerNode)(Arc);
Factory_1.Factory.addGetterSetter(Arc, 'innerRadius', 0, (0, Validators_1.getNumberValidator)());
Factory_1.Factory.addGetterSetter(Arc, 'outerRadius', 0, (0, Validators_1.getNumberValidator)());
Factory_1.Factory.addGetterSetter(Arc, 'angle', 0, (0, Validators_1.getNumberValidator)());
Factory_1.Factory.addGetterSetter(Arc, 'clockwise', false, (0, Validators_1.getBooleanValidator)());
